cfme/tests/openstack/cloud/test_volumes.py

```diff
@@ -0,0 +1,202 @@
+    for volume in provider.mgmt.list_volume():
+        vname = provider.mgmt.get_volume(volume).to_dict()['display_name']
+        volumes.append(vname)
+    provider.load_details()
+    err_msg = 'Details page shows wrong number of provider volumes'
+    clv_el = provider.summary.relationships.cloud_volumes
+    assert len(volumes) == clv_el.value, err_msg
+
+    clv_el.click()
+    err_msg = 'One of volumes is not displayed'
+    for volume in volumes:
+        assert list_tbl.find_cell('Name', volume), err_msg
+
+
+@pytest.mark.uncollectif(lambda: PROD_VERSION > '4.1')
+def test_attach_volume_from_instance_page(provider):
+    """Attaches pre-created volume to an instance from Instance page"""
+    vms = filter(lambda vm: vm.power_state == 'ACTIVE', provider.mgmt.all_vms())
+    instance_name = choice(vms).name
+    volume_name = None
+    # Find available volume
+    for volume in provider.mgmt.list_volume():
+        vol_info = provider.mgmt.get_volume(volume).to_dict()
+        if vol_info['status'] == 'available':
+            volume_name = vol_info['display_name']
+            break
+    assert volume_name, "Can't find free volume to attach"
+
+    Instance(instance_name, provider).load_details()
+    toolbar.select('Configuration', 'Attach a Cloud Volume to this Instance')
+    select = Select("//select[@id='volume_id']")
+    assert len(select.all_options) > 0
+
+    select.select_by_value(select.get_value_by_text(volume_name))
+    pytest_selenium.send_keys("//input[@name='device_path']", '/dev/vdb')
+    pytest_selenium.click("//button[@id='save_enabled']")
+    flash_msg_loc = "//div[@id='flash_text_div']/div/strong"
+    msg = pytest_selenium.text(flash_msg_loc)
+    assert 'Attaching Cloud Volume' in msg
+    pytest_selenium.click(flash_msg_loc)
+
+    wait_for(provider.is_refreshed, [None, 10], delay=5)
+    info_el = InfoBlock.element('Relationships', 'Cloud Volumes')
+    pytest_selenium.click(info_el)
+    params = {'Name': volume_name, 'Status': 'in-use'}
+    res = list_tbl.find_rows_by_cells(params)
+    assert res, 'Volume does not appear in instance relationships'
+
+
+@pytest.mark.uncollectif(lambda: PROD_VERSION > '4.1')
+def test_detach_volume_from_instance_page(provider):
+    """Detaches volume from instance from Instance page"""
+    vm_name = None
+    # Find an instance with attached volumes
+    for instance in provider.mgmt._get_all_instances():
+        info = instance.to_dict()
+        if info['os-extended-volumes:volumes_attached']:
+            vm_name = info['name']
+            break
+    assert vm_name, "Can't find any instance with attached volume"
+
+    Instance(vm_name, provider).load_details()
+    toolbar.select('Configuration', 'Detach a Cloud Volume from this Instance')
+    select = Select("//select[@id='volume_id']")
+    v_option = choice(select.all_options[1:])
+    select.select_by_value(v_option[1])
+    loc = "//div[@id='angular_paging_div_buttons']/button[@id='save_enabled']"
```

Comment:
Created missing form buttons and cloud volumes form etc. Replaced all hardcoded elements with them